src/reify.cpp

```diff
@@ -0,0 +1,333 @@
+
+    void clear() {
+        theoryTuples.clear();
+        theoryElementTuples.clear();
+        litTuples.clear();
+        atomTuples.clear();
+        weightLitTuples.clear();
+        nodes.clear();
+        graph.clear();
+    }
+};
+
+Reifier::Reifier(std::ostream& out, const Options& opts)
+    : out_(out)
+    , calculateSccs_(opts.calculateSccs)
+    , reifyStep_(opts.reifyStep)
+    , stepData_(std::make_unique<StepData>()) {}
+
+Reifier::~Reifier() noexcept = default;
+
+template <typename... T>
+void Reifier::printFact(const char* name, const T&... args) {
+    out_ << name << "(";
+    printCommaSeparated(out_, args...);
+    out_ << ").\n";
+}
+
+template <typename... T>
+void Reifier::printStepFact(const char* name, const T&... args) {
+    if (reifyStep_) {
+        printFact(name, args..., step_);
+    }
+    else {
+        printFact(name, args...);
+    }
+}
+
+template <typename M, typename T>
+auto Reifier::tuple(M& map, const char* name, const T& args) -> size_t {
+    return tuple(map, name, toVec(args));
+}
+
+template <typename M, typename T>
+auto Reifier::tuple(M& map, const char* name, std::vector<T>&& args) -> size_t {
+    std::sort(args.begin(), args.end());
+    auto [it, inserted] = map.emplace(std::move(args), map.size());
+    if (inserted) {
+        printStepFact(name, it->second);
+        for (const auto& x : it->first) { printStepFact(name, it->second, x); }
+    }
+    return it->second;
+}
+
+template <typename M, typename T>
+auto Reifier::orderedTuple(M& map, const char* name, const T& args) -> size_t {
+    return orderedTuple(map, name, toVec(args));
+}
+
+template <typename M, typename T>
+auto Reifier::orderedTuple(M& map, const char* name, std::vector<T>&& args) -> size_t {
+    auto [it, inserted] = map.emplace(std::move(args), map.size());
+    if (inserted) {
+        printStepFact(name, it->second);
+        int arg = 0;
+        for (const auto& x : it->first) {
+            printStepFact(name, it->second, arg, x);
+            ++arg;
+        }
+    }
+    return it->second;
+}
+
+auto Reifier::theoryTuple(IdSpan args) -> size_t { return orderedTuple(stepData_->theoryTuples, "theory_tuple", args); }
+
+auto Reifier::theoryElementTuple(IdSpan args) -> size_t {
+    return tuple(stepData_->theoryElementTuples, "theory_element_tuple", args);
+}
+
+auto Reifier::litTuple(LitSpan args) -> size_t { return tuple(stepData_->litTuples, "literal_tuple", args); }
+
+auto Reifier::weightLitTuple(WeightLitSpan args) -> size_t {
+    return tuple(stepData_->weightLitTuples, "weighted_literal_tuple", args);
+}
+
+auto Reifier::atomTuple(AtomSpan args) -> size_t { return tuple(stepData_->atomTuples, "atom_tuple", args); }
+
+auto Reifier::addNode(Atom_t atom) -> uint32_t {
+    auto& nodesMap = stepData_->nodes;
+    auto  it       = nodesMap.find(atom);
+    if (it != nodesMap.end()) {
+        return it->second;
+    }
+    auto nodeId    = stepData_->graph.addNode(atom);
+    nodesMap[atom] = nodeId;
```

Comment:
This is an unnecessary second map lookup. Consider simplifying to something like:
```cpp
auto Reifier::addNode(Atom_t atom) -> uint32_t {
    auto [it, isNew] = stepData_->nodes.try_emplace(atom, 0);
    if (isNew) {
        it->second = stepData_->graph.addNode(atom);
    }
    return it->second;
}
```